signing_service/src/signing_service.py

```diff
@@ -0,0 +1,167 @@
+        assert isinstance(port, int)
+        assert isinstance(initial_reconnect_delay, int)
+        self.host = host
+        self.port = port
+        self.initial_reconnect_delay = initial_reconnect_delay
+        self.current_reconnect_delay = None
+        self.socket: socket.socket = None  # type: ignore
+        self.was_sigterm_caught = False
+
+    def run(self) -> None:
+        """
+        Handles main connection loop and its exceptions.
+
+        If connection is interrupted due to a failure, it should try to reconnect.
+        If connection is closed by other side, exit gracefully.
+        If a shutdown signal or KeyboardInterrupt is caught, exit gracefully.
+        If there was an unrecognized exception, it logs it and report to Sentry, then reraise and crash.
+        """
+
+        def _set_was_sigterm_caught_true(signum, frame):  # pylint: disable=unused-argument
+            self.was_sigterm_caught = True
+
+        # Handle shutdown signal.
+        signal.signal(signal.SIGTERM, _set_was_sigterm_caught_true)
+
+        while not self.was_sigterm_caught:
+            try:
+                if self.socket is None:
+                    self._connect()
+                else:
+                    # Sending empty string because it is only way to check
+                    # if connection is still alive directly on socket object.
+                    self.socket.send(b'')
+            except socket.error as exception:
+                logger.error(f'Socket error occurred: {exception}')
+
+                # When the other side closes the connection, return without errors.
+                if isinstance(exception.args, tuple) and exception.args[0] == socket.errno.EPIPE:  # type: ignore
+                    logger.info(os.strerror(exception.args[0]))
+                    break
+
+                # Only predefined list of exceptions should cause reconnect, others should be reraised.
+                if isinstance(exception.args, tuple) and exception.args[0] not in SIGNING_SERVICE_EXCEPTIONS_CAUGHT_TO_RECONNECT:  # type: ignore
+                    raise
+
+                # Increase delay and reconnect if the connection is interrupted due to a failure.
+                self._increase_delay()
+                self.socket = None
+            except KeyboardInterrupt:
+                # Handle keyboard interrupt.
+                logger.info('Closing connection and exiting on KeyboardInterrupt.')
+                break
+            except Exception as exception:
+                # If there was an unrecognized exception, log it and report to Sentry.
+                crash_logger.error(f'Unrecognized exception occurred: {exception}')
+                raise
+            finally:
+                if self.socket is not None:
+                    # Try to close opened socket.
+                    self.socket.close()
+
+        if self.was_sigterm_caught:
+            if self.socket is not None:
+                self.socket.close()
```

Comment:
You would not need this if you put the whole loop in the `finally` block.

Reply:
I am not sure if I follow you on this one. This is actually part of the loop.

Reply:
I mean that the the whole loop should be inside a `try...finally` block. You should start such a block immediately after acquiring a resource. If you did this, this `close()` would not be necessary. The block would handle that for you.

And now that I think of it - this is really a place where you should use a context manager. Then the finally block would not be even necessary. Just like you do `with open(...):`. With socket it would look like this:

``` python
from contextlib import closing

with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as tcp_socket:
```

See [How to use socket in Python as a context manager?](https://stackoverflow.com/questions/16772465/how-to-use-socket-in-python-as-a-context-manager/16772520).

This also shows what I meant in my comment about passing `socket` in parameters, not on `self`: the `with` block should be the scope for the socket and it should be destroyed automatically when the scope ends. Putting it on `self` may potentially extend its life past it. The `with` block guarantees that it won't happen, even by mistake.

Reply:
The context manager looks nice but I still don't understand how you want this code to reconnect on several errors caught in `try..except` block if loop will be inside `try`. The only option I can think is adding another `try..except` inside `try`. Is that what you have meant ?

Reply:
OK, sorry. `with` should be inside the loop.

Reply:
Ok, I have rebuild it and simplify a little bit :+1: done